### DISCUSSION TITLE
Injection grammar extension

### DISCUSSION THREAD

**Discussion by ravirathore** [microsoft/vscode-discussions#152] [Extension Development QnA] [answered]
- VS Code Version: 1.71.2
- OS Version: Windows 10

Steps to Reproduce:

Hi,
I wanted to build one small injection grammar extension. 

I followed this tutorial. https://code.visualstudio.com/api/language-extensions/syntax-highlight-guide.

And, updated these 2 files:
1. package.json
2. syntaxes/injection.json

I tested it for javascript, python and julia languages; respectively updating files with 'source.js', 'source.python' and 'source.julia'. Results are in the end.

Results(in debugging mode) are: 
1. It worked for julia. .
2. But, not in python and javascript

Please let know if I miss anything. 

Please advise.

Thanks and regards,

**Comment by RedCMD** [comment]
I assume javascripts semantic highlighting overrides all textmate highlighting

As far as I know
there is no realistic way of getting around this
other than rewriting large portions of javascripts extension

**Comment by ravirathore** [comment]
Thanks for the reply.

Do we have to build a semantic token provider, which can find and send back tokens.

Regards,

**Comment by RedCMD** [comment]
only one semantic token provider can be active per document
and there is no way to inject into javascripts one

the provider does not give you tokens
you give it the tokens that you have parsed with your own parser

**Comment by ravirathore** [comment]
Thanks.

Sorry, understood.


regards,

**Comment by ravirathore** [comment]
In python code, after disabling semantic highlighting and python extension, its still not highlighting ..

Please advise.

**Comment by RedCMD** [comment, marked as answer]
You cannot inject directly into an already captured token (content of `"begin"`/`"end"` rules are different)
you have to replace and handle the token you're trying to modify
https://github.com/MagicStack/MagicPython/blob/2802ded681e0ab1a1057821c1da287147d639505/grammars/MagicPython.tmLanguage#L619-L623
![image](https://user-images.githubusercontent.com/33529441/194020973-bd961aea-51c0-4819-934c-aa10ed5387a6.png)
Pythons `variable` capture is easy to copy: `\\b[[:alpha:]_]\\w*\\b`

Julia doesn't capture its variables

**Comment by ravirathore** [comment]
Thanks
